docs/painless/painless-guide/painless-datetime.asciidoc

```diff
@@ -0,0 +1,295 @@
+int nanos = 0;
+String tz = 'Z';
+ZonedDateTime zdt = ZonedDateTime.of(
+        year, month, day, hour, minutes, seconds, nanos, ZoneId.of(tz));
+----
+
+===== Datetime Formatting Examples
+
+* format to a long (milliseconds)
++
+[source,Painless]
+----
+ZonedDateTime zdt =
+        ZonedDateTime.of(1983, 10, 13, 22, 15, 30, 0, ZoneId.of('Z'));
+long milliSinceEpoch = zdt.toInstant().toEpochMilli();
+----
++
+* format to a String (ISO 1806)
++
+[source,Painless]
+----
+ZonedDateTime zdt =
+        ZonedDateTime.of(1983, 10, 13, 22, 15, 30, 0, ZoneId.of('Z'));
+String datetime = zdt.format(DateTimeFormatter.ISO_INSTANT);
+----
+Note the specification of a built-in DateTimeFormatter.
++
+* format to a String (custom)
++
+[source,Painless]
+----
+ZonedDateTime zdt =
+        ZonedDateTime.of(1983, 10, 13, 22, 15, 30, 0, ZoneId.of('Z'));
+DateTimeFormatter dtf = DateTimeFormatter.ofPattern(
+        "'date:' yyyy/MM/dd 'time:' HH:mm:ss");
+String datetime = zdt.format(dtf);
+----
+Note the specification of a custom DateTimeFormatter.
++
+* format to individual pieces
++
+[source,Painless]
+----
+ZonedDateTime zdt =
+        ZonedDateTime.of(1983, 10, 13, 22, 15, 30, 100, ZoneId.of(tz));
+int year = zdt.year;
+int month = zdt.monthValue;
+int day = zdt.dayOfMonth;
+int hour = zdt.hour;
+int minutes = zdt.minute;
+int seconds = zdt.second;
+int nanos = zdt.nano;
+----
+
+==== Datetime Modification
+
+Modifications to datetimes such as adding/subtracting a few seconds or setting
+a specific day are often required to meet a script's objectives. Use either a
+long or a ZonedDateTime to do datetime modification. Use standard
+<<painless-operators-numeric, numeric operators>> to modify a long
+representation of a datetime. Use <<method-call-operator, method calls>> to
+modify a ZonedDateTime representation of a datetime. Note most modification
+methods for a ZonedDateTime return a new instance that needs assignment or
+immediate use.
+
+===== Datetime Modification Examples
+
+* Subtract three seconds from a long (milliseconds)
++
+[source,Painless]
+----
+long milliSinceEpoch = 434931330000L;
+milliSinceEpoch = milliSinceEpoch - 1000L*3L;
+----
++
+* Add three days to a datetime
++
+[source,Painless]
+----
+ZonedDateTime zdt =
+        ZonedDateTime.of(1983, 10, 13, 22, 15, 30, 0, ZoneId.of('Z'));
+ZonedDateTime updatedZdt = zdt.plusDays(3);
+----
++
+* Subtract 125 minutes from a datetime
++
+[source,Painless]
+----
+ZonedDateTime zdt =
+        ZonedDateTime.of(1983, 10, 13, 22, 15, 30, 0, ZoneId.of('Z'));
+ZonedDateTime updatedZdt = zdt.minusMinutes(125);
+----
++
+* Set the year on a datetime
++
+[source,Painless]
+----
+ZonedDateTime zdt =
+        ZonedDateTime.of(1983, 10, 13, 22, 15, 30, 0, ZoneId.of('Z'));
+ZonedDateTime updatedZdt = zdt.withYear(1976);
+----
+
+==== Elapsed Time
+
+A script's objectives may require the elapsed time (difference) between two
+datetimes. Use either two longs or two ZonedDateTimes to calculate an elapsed
```

Comment:
Maybe we could just explain here that epoch millis in longs can be compared for a difference in millis, but for more complex comparisons ZonedDateTime should be used.

Reply:
Cleaned up a bit.